logstash-core/src/main/java/org/logstash/execution/AbstractPipelineExt.java

```diff
@@ -593,6 +583,71 @@ private UptimeMetric initOrGetUptimeMetric(final ThreadContext context,
+            IRubyObject config = rubyObject.callMethod(context, "config");
+            String id = config.convertToHash().get("id").toString();
+            initializePluginThroughputFlowMetric(context, uptimeInPreciseMillis, id);
+        }
+
+        final int workerCount = getSetting(context, SettingKeyDefinitions.PIPELINE_WORKERS).convertToInteger().getIntValue();
+
+        for (AbstractFilterDelegatorExt delegator: lirExecution.filters()) {
+            initializePluginWorkerFlowMetrics(context, workerCount, uptimeInPreciseMillis, FILTERS_KEY, delegator.getId().asJavaString());
+        }
+
+        for (AbstractOutputDelegatorExt delegator: lirExecution.outputs()) {
+            initializePluginWorkerFlowMetrics(context, workerCount, uptimeInPreciseMillis, OUTPUTS_KEY, delegator.getId().asJavaString());
+        }
+    }
+
+    private void initializePluginThroughputFlowMetric(final ThreadContext context, final Metric<Number> uptimeInPreciseMillis, final String id) {
+        final RubySymbol[] eventsNamespace = buildNamespace(PLUGINS_KEY, INPUTS_KEY, RubyUtil.RUBY.newSymbol(id), EVENTS_KEY);
+        final LongCounter counter = initOrGetCounterMetric(context, eventsNamespace, OUT_KEY);
+
+        final FlowMetric throughputFlow = createFlowMetric(PLUGIN_THROUGHPUT_KEY, counter, uptimeInPreciseMillis);
+        this.flowMetrics.add(throughputFlow);
+
+        final RubySymbol[] flowNamespace = buildNamespace(PLUGINS_KEY, INPUTS_KEY, RubyUtil.RUBY.newSymbol(id), FLOW_KEY);
+        storeMetric(context, flowNamespace, throughputFlow);
+    }
+
+    private void initializePluginWorkerFlowMetrics(final ThreadContext context, final int workerCount, final Metric<Number> uptimeInPreciseMillis, final RubySymbol key, final String id) {
+        final RubySymbol[] eventsNamespace = buildNamespace(PLUGINS_KEY, key, RubyUtil.RUBY.newSymbol(id), EVENTS_KEY);
+        final LongCounter durationInMillis = initOrGetCounterMetric(context, eventsNamespace, DURATION_IN_MILLIS_KEY);
+        final LongCounter counterEvents = initOrGetCounterMetric(context, eventsNamespace, IN_KEY);
+        final FlowMetric workerCostPerEvent = createFlowMetric(WORKER_COST_PER_EVENT_KEY, durationInMillis, counterEvents);
+        this.flowMetrics.add(workerCostPerEvent);
+
+        final UpScaledMetric percentScaledDurationInMillis = new UpScaledMetric(durationInMillis, 100);
+        final UpScaledMetric availableWorkerTimeInMillis = new UpScaledMetric(uptimeInPreciseMillis, workerCount);
+        final FlowMetric workerUtilization = createFlowMetric(WORKER_UTILIZATION_KEY, percentScaledDurationInMillis, availableWorkerTimeInMillis);
+        this.flowMetrics.add(workerUtilization);
+
+        final RubySymbol[] flowNamespace = buildNamespace(PLUGINS_KEY, key, RubyUtil.RUBY.newSymbol(id), FLOW_KEY);
+        storeMetric(context, flowNamespace, workerCostPerEvent);
+        storeMetric(context, flowNamespace, workerUtilization);
+    }
```

Comment:
Thoughts:

I think it makes sense to get this in largely as-is, but to follow up with a migration to the plugin's `AbstractFilterDelegatorExt#initMetrics` and `AbstractOutputDelegatorExt#initMetrics`(which already do a lot of this work and memoize their own references to the underlying metrics), and possibly finding a way to unify delegation to java- and ruby-based inputs and codecs. Doing so would require having a channel back to the `flowMetrics` list so that they could be reached by our periodic poller.

Reply:
It would be an exception for the inputs since we don't have similar structure with filters & outputs but I do agree that moving to the relative scope would be better. Let me check if I can quickly re-organize, at least will follow up after delivering.

Reply:
As much as I try, majority sources are going to be duplicated: (`AbstractPipilineExt` classes') `initOrGetUptimeMetric` to get uptime metric and `storeMetric` to store the metrics. Extra requirements: to share the worker count through the context and store flow metrics independently for each plugin. We will also need to make a `collect` interface for plugins as we are having for pipeline.
In my opinion, from the domain perspective and to make flow metric isolated, it would be better to separate flow metric sources into package. So that it will not be any conflict boundaries with plugins/pipelines.

Reply:
I've got a refactor idea that may clean this allllll up nicely, but I'd rather merge this with working functionality before tackling.